gcov plugin - Support for gcovr 4.2 and additional gcovr options

aschlicht opened the pull request:
The gcov plugin detects which version of gcovr is available to call gcovr with the correct command line options. Between 4.1 and 4.2, support for SonarQube and JSON reports were added. Also added was the ability to generate multiple reports from one call to gcovr. When the gcov plugin detects version <= 4.1, the gcovr will be invoked twice if XML and HTML output is enabled. When the gcov plugin detects version >= 4.2, then all command line parameters will be joined for a single call to gcovr.

Support for nearly all gcovr command line options was added. We can specify them with a similarly named option in project.yml. I copied the help text from the gcovr user guide and added a citation to reference their work. Don't want to infringe on a copyright. We are are trying to support use of their tool!

aschlicht commented on lib/ceedling/file_path_utils.rb:
```diff
@@ -21,9 +21,11 @@ class FilePathUtils
 
   # standardize path to use '/' path separator & have no trailing path separator
   def self.standardize(path)
-    path.strip!
```
The `html_absolute_paths` option caused an exception when project.yml was parsed because the option is boolean. I changed this function to replace \ with / only if the path is a string.

aschlicht commented:
I see the build failed because tests were looking for specific messages printed to the console. Do you want to keep the original messages? I changed the messages because multiple reports may be generated from a single call to gcovr. I made one generic message for all report types.

aschlicht commented:
I spent time adding all the gcovr options. I suppose we could abandon this approach and give users a single project.yml option to directly specify command line options when calling gcovr.

```yaml
:gcov:
    :command_line_options: <custom_command_line_options>
```

For example

```yaml
:gcov:
    :command_line_options: --branches --exclude-unreachable-branches
```

CezaryGapinski commented:
Hi!
How to use :print_summary: option? I set
```
:gcov:
  :print_summary: TRUE
```
 in my project.yml and I can't see any additional summary in console.

aschlicht commented:
@CezaryGapinski Thank you for trying the latest changes. You are correctly using the option. I also did not see any additional output with the existing code, so I fixed the problem. Please get the latest and try again.

I also just added the `:text_report: [TRUE|FALSE]` and `:text_artifact_filename: <output>` options. The text report is disabled by default. If you specify `:text_report: true`, then gcovr will be called an additional time without the other html, xml, json report options. The text report will be output to a file if you specify the `:text_artifact_filename:`, otherwise it will be printed to the console. You may choose lines or branches metrics in the text report with the `:branches:` option.

```
Creating a gcov text report... Done in 1.327 seconds.
lines: 68.2% (307 out of 450)
branches: 42.0% (42 out of 100)
```

I just added the execution time too.

CezaryGapinski commented:
@aschlicht Now text report is working fine. I didn't try all implemented features yet, but for me this PR looks really great.
I have started implementation with add json tracefiles feature to merge results for sources compiled with different options (for example different settings of preprocessor defines). The text report from gcovr is really helpful. Report from gcov do not merge results and is useless in this case.

CezaryGapinski commented on plugins/gcov/gcov.rake:
```diff
@@ -154,67 +154,266 @@ if PROJECT_USE_DEEP_DEPENDENCIES
+
+  # Run gcovr with the given arguments.
+  def run_gcovr(args)
+    command = @ceedling[:tool_executor].build_command_line(TOOLS_GCOV_POST_REPORT, [], args)
```
Here to run gcovr you are using executable from gcov_post_report defined in yml file.
To detect gcovr version you are using `gcovr_version_info = %x|gcovr --version|`
It means if I set:
```
:gcov_post_report:
  :executable: my_custom_gcovr
```
the version will be detected from incorrect gcovr?

aschlicht replied:
@CezaryGapinski Thank you for your continued thorough review. You are correct that I need to fix the way the gcovr version is detected. Let's not yet complete the pull request.

aschlicht replied:
@CezaryGapinski  I pushed changes to fix the problem you found with getting the version info. What other issues can you find?

mvandervoord commented:
Thank you both for your continued support! When you are both feeling comfortable, can you ping me with a directed comment so I can merge?

aschlicht commented:
@mvandervoord Thank you, Mark!

I've been playing with creating reports from Report Generator with the gcov plugin. I like the look of Report Generator reports a bit better (personal preference). They show yellow for partially covered lines while gcovr html reports show either green or red. Report Generator also has the ability to link in a history file to show coverage over time.

I added Report Generator features to the gcov plugin because Report Generator can take gcov or Cobertura as an input (among many others). At the moment, I found it very easy to have gcovr grab all the gcov data and combine it into a Cobertura xml report. I then run Report Generator with the Cobertura xml data.

I'd like your thoughts regarding putting Report Generator features in a new plugin versus including in the gcov plugin.

https://github.com/danielpalme/ReportGenerator#usage

```yaml
:gcov:
    :report_generator: true
    :report_generator_report_types: Html;Latex
```

```
Creating gcov results report(s) in 'build/artifacts/gcov'... Done in 0.936 seconds.
Creating gcov results report(s) with Report Generator in 'build/artifacts/gcov/ReportGenerator'... Done in 0.605 seconds.
```

I have not yet committed or pushed the report generator features, so this pull request does not include them. I can put the changes into a branch on my fork if you want to see.

I suggest installing Report Generator as a global tool. There are other ways to install it.
https://www.nuget.org/packages/dotnet-reportgenerator-globaltool

aschlicht commented:
@mvandervoord @CezaryGapinski 

The current solution uses boolean options to enable or disable report outputs.

```yaml
:gcov:
    :html_report: true
    :xml_report: true # Same output as cobertura_report.
    :cobertura_report: true
    :sonarqube_report: true
    :json_report: true
    :text_report: true
```

What do you think about a single option to select report types?

```yaml
:gcov:
    :report_types: GcovrHtml;GcovrHtmlSummary;Cobertura;SonarQube;JSON;GcovrText;ReportGeneratorHtml
```

Do you like the separate boolean options or the combined option better?

mvandervoord commented:
How about making it a list like other features in Ceedling?

```
:gcov:
    :report_types: 
      - GcovrHtml
      - GcovrHtmlSummary
      - Cobertura
      - SonarQube
      - JSON
      - GcovrText
      - ReportGeneratorHtml
```

aschlicht commented:
@mvandervoord Good idea. I like consistency and it makes sense to use the the yaml list type. I'll make the change.

I think the `html_report` and `xml_report` options have been available in the ceedling gem since v0.29.0 released this month. Do we need to keep them for backward compatibility? We could keep `html_report_type` and put a documentation note or warning console output about being deprecated.

```yaml
:gcov:
    :report_types: 
      - GcovrHtmlBasic # <-- Shall we keep the Basic and Detailed nomenclature?
      - GcovrHtmlDetailed
      - Cobertura
      - SonarQube
      - JSON
      - GcovrText
      - ReportGeneratorHtml
```

I put Gcovr in the HTML names in case different HTML report generators are supported. I think Cobertura, SonarQube, and JSON gcovr outputs follow external standards.

mvandervoord commented:
Having `xml_report` and `html_report report` as deprecated but still work would be good. While 0.29.0 was the first time it was officially released, people have been using it for some time as cloned directly off github. (my fault for not doing releases often enough).

I enjoy `basic` vs `detailed`. I think it's intuitive what it means.

CezaryGapinski commented on plugins/gcov/gcov.rake:
```diff
@@ -154,67 +154,268 @@ if PROJECT_USE_DEEP_DEPENDENCIES
+    args += "--root \"#{opts[:gcov_report_root] || '.'}\" "
+    args += "--config \"#{opts[:gcov_config_file]}\" " unless opts[:gcov_config_file].nil?
+    args += "--filter \"#{opts[:gcov_report_include]}\" " unless opts[:gcov_report_include].nil?
+    args += "--exclude \"#{opts[:gcov_report_exclude] || GCOV_FILTER_EXCLUDE}\" "
```
@aschlicht I know you are working on changes for `:report_types:` list. I'm looking forward to see this new version :).

I just have a question about these options and description in README.md

```
# Keep only source files that match this filter. (gcovr --filter).
:report_include: "^src"
 
# Exclude source files that match this filter. (gcovr --exclude).
:report_exclude: "^build|^vendor|^test|^support"
```
`GCOV_FILTER_EXCLUDE` has value `'^vendor.*|^build.*|^test.*|^lib.*'` and is always ORed with `gcov_report_exclude`

`gcov_report_include` is not ORed with any settings but in the README.md you mentioned about filter for `^src`

Settings mentioned in the README.md are related to the default behaviour or it is just an example how to use these options?

aschlicht replied:
@CezaryGapinski I was treating the settings in the README.md as examples and documenting some defaults in the comments above the settings. I can update the `:report_exclude:` in README.md to show the default.

Regarding the OR, I thought the `||` was a logical OR. So if project.yml options specify `:report_exclude:`, then use the settings in project.yml. If the settings are not included in the options, then use the default specified by `GCOV_FILTER_EXCLUDE`. I read about this Ruby syntax on the internet. Did I interpret it incorrectly?

I have some of the `:report_types:` version implemented. Shall we support both GcovrHtmlBasic and GcovrHtmlDetailed? We could output only detailed in that case.

CezaryGapinski replied:
> Regarding the OR, I thought the || was a logical OR. So if project.yml options specify :report_exclude:, then use the settings in project.yml. If the settings are not included in the options, then use the default specified by GCOV_FILTER_EXCLUDE. I read about this Ruby syntax on the internet. Did I interpret it incorrectly?

Now I understand it is clear for me now. I guess you have right about || operator.
I just was not sure about difference between `'^vendor.*|^build.*|^test.*|^lib.*'` and `"^build|^vendor|^test|^support"`.

aschlicht replied:
Thanks for confirming the use of the || operator.

I changed the default in the README.md as you suggested.

CezaryGapinski commented:
> ```yaml
> :gcov:
>     :report_types: 
>       - GcovrHtmlBasic # <-- Shall we keep the Basic and Detailed nomenclature?
>       - GcovrHtmlDetailed
>       - Cobertura
>       - SonarQube
>       - JSON
>       - GcovrText
>       - ReportGeneratorHtml
> ```

When ReportGeneratorHtml is set, then basic report will be the default behaviour and GcovrHtmlBasic is redundant? IMHO then could be removed.

aschlicht commented:
I added the Gcovr prefix to HtmlBasic and HtmlDetailed in case we want to support html reports from other tools. The Gcovr prefix gives us a way to say, "I want an HTML report from gcovr." Or, "I want an HTML report from Report Generator." If we want to keep Report Generator out of this plugin, then we could remove the Gcovr prefix. I like to throw around many ideas to see which ones make sense.

I pushed changes to support the `:report_types:` option. Please take a look when you have time. Thanks!

CezaryGapinski commented:
Ok. Maybe is better to add new section named utils and remove prefixes from the report types?

:gcov:
  :utilities:
    - gcovr
    - ReportGenerator
  :report_types:
    - HtmlBasic
    - HtmlDetailed
    - Cobertura
    - SonarQube
    - JSON
    - Text

You can mention in the README.md about which of the report types are availiable for the specified utility.

aschlicht commented:
@CezaryGapinski I like your `:utilities:` idea better than my `Gcovr` prefix idea. We can default to gcovr when `:utilities:` is unspecified.

CezaryGapinski commented on plugins/gcov/README.md:
```diff
@@ -23,79 +23,239 @@ pip install gcovr
 
-Finally, there are a number of settings which can be specified in order to adjust the
+Finally, there are a number of settings which can be specified to adjust the
 default behaviors of gcov:
```
I think these options specify behaviours of gcovr not gcov. For now it is ok, because we only have the gcovr, but after extend for the ReportGenerator it may not be true.

aschlicht replied:
Updated `gcov` to `gcovr` as requested.

aschlicht replied:
Do we want to put all gcovr specific options under a :gcovr: section?

```yaml
:gcov:
    :gcovr:
        :report_root: "."
        :keep: true
    :report_generator:
        :verbosity: Info
```

CezaryGapinski commented on plugins/gcov/gcov.rake:
```diff
@@ -154,67 +154,310 @@ if PROJECT_USE_DEEP_DEPENDENCIES
-      command = @ceedling[:tool_executor].build_command_line(TOOLS_GCOV_POST_REPORT_XML, [], args)
-      @ceedling[:tool_executor].exec(command[:line], command[:options])
+    if is_utility_enabled(opts, UTILITY_NAME_GCOVR)
+      make_gcovr_reports(opts)
```
If here we have the condition to check if enabled utility is gcovr, so other is_utility_enabled(opts, UTILITY_NAME_GCOVR) calls in internal gcovr_*() methods are not redundant?

aschlicht replied:
I removed the nested / redundant function calls. I thought about this before committing the code and left it. Since you pointed out the issue, it's worth removing to simplify the inner logic.

CezaryGapinski commented:
@aschlicht @mvandervoord 
For me this Pull Request looks ok. I have found only small problems. All of these problems are already fixed. @aschlicht made a great job. His changes allow to use almost all features of gcovr.
Gcov plugin can be also enhanced for the ReportGenerator coverage report in the future.

mvandervoord commented:
@aschlicht -- (A) I just want to verify you are happy too and (B) THANK YOU for all this work!

aschlicht commented:
@mvandervoord You're welcome! Thanks for the, "Thank you". I'm glad you appreciate the work. Thank you for promptly chiming in with configuration interface guidelines. I do want to match the look and feel of Ceedling.

I am happy with the pull request; however, I have a couple more questions. If it makes sense to add Report Generator output to the gcov plugin, then do we want to separate gcovr options from Report Generator options in the project.yml hierarchy?

1. Do we want to put all gcovr specific options under a :gcovr: section?
1.I think I should rename `:report_types:` to simply `:reports:`. Do you have a better name?

```yaml
:gcov:
    :report_types: # <--- Rename?
        - HtmlDetailed
        - JSON
    :gcovr:
        :report_root: "."
        :html_medium_threshold: 75
        :keep: true
    :report_generator:
        :history_directory: "../history"
        :verbosity: Info
        :file_filters:
            - "-^test_.*"
            - "+..\/..\/Application\/.*"
```

mvandervoord commented:
I think those changes make a lot of sense. I like the organization and I like the simplicity of naming it just `:report`

I also think it would make sense to change to that format in this PR, even if we're adding report_generator later, to reduce the number of times people need to update their configuration. :)

aschlicht commented:
@CezaryGapinski Thank you for your testing, reviewing, and support for this pull request. You mentioned you had ideas for supporting JSON trace files (--add-tracefile). I don't know your thoughts, but I have an idea to suggest after Mark suggested the yaml list for specifying reports outputs.

```yaml
:gcov:
    :tracefiles:
        - filename1.json
        - filename2.json
        - filename3.json
```

Within gcov.rake, we would loop on the array. For each item, we call:

```ruby
args += "--add-tracefile #{filename} "
```

This works for known trace files. Maybe you have a good idea for adding dynamic trace file names and quantities? The array of filenames could be an array of regex instead of simply strings to glob multiple files.

aschlicht commented:
@mvandervoord I agree we should add :gcovr: hierarchy now. Are you ok with me putting these recent additions under the :gcovr: hierarchy? That would be a non-backward compatible change.

```yaml
:gcov:
  :html_medium_threshold: 75
  :html_high_threshold: 90
  :fail_under_line: 30
  :fail_under_branch: 30
```

CezaryGapinski commented:
@aschlicht No problem at all :). You did a lot of work! Thank you!

Going back to your question, I was thinking about merging results in dynamic way.

When you compile source file then .gcno and .gcda files are created. When you run the executable then the .gcda file contains counts of execution of lines, branches, etc.
When you compile your source file again for different options (other preprocessor definitions) .gcno and .gcda are created again, and old results are overwritten and are not available. You can find an example in [gcovr add coverages test](https://github.com/gcovr/gcovr/tree/master/gcovr/tests/add_coverages) and you will know what exactly I have in mind.

In ceedling I would like to store previous results in JSON file created by gcovr to merge it with the new results.

I just thinking about add the option merge_results.

```yaml
:gcov:
  :gcovr:
    :merge_results: [true, false]
```

Temporary JSON files will be placed in `build/gcov/out` and eventually other reports placed in `:reports:` will be created. That was my plans when I was implementing JSON feature in gcovr.

You idea about `--add-tracefile` looks ok, but I can't find practical usage of:
```yml
:gcov:
    :tracefiles:
        - filename1.json
        - filename2.json
        - filename3.json
```
Do you have an idea for what use filename[1,3].json in practice :)?

aschlicht commented:
@CezaryGapinski I didn't know you added the JSON feature to gcvor. Cool! Note that when I used ---json <filename> without --output <filename>, then no JSON file was created. The gcovr documentation says --json <filename> is supported. `gcovr --json coverage.json`

Back to Ceedling... Your merge idea sounds good. Right now the JSON report is output to `build/artifacts/gcov`. Does that conflict with your merge idea?

I was thinking that the JSON file input support was to combine coverage information from other portions of project to get a complete picture of coverage results. I didn't think it through though. What if we have a collection of git submodules, each with its own set of unit tests? We combine a set of submodules to create a product. We want to know the unit test coverage for the product, so we specify a set of JSON files to import. Maybe we directly call gcovr for this approach?

aschlicht commented:
I moved all the gcovr settings under `:gcov:gcovr:`. I did some testing and I'm pretty sure I got them all. I made it easier to change the setting path in the future.

mvandervoord commented:
sorry about the delay -- yes I think it's fine to make changes to the way we're configuring this plugin. If we're going to break compatibility, doing it all at once is the way to go.

aschlicht commented:
@mvandervoord @CezaryGapinski I've made all the changes and fixes we discussed. I'm ready to complete the pull request as long as you are both happy with it as well. Feel free to examine the README too: https://github.com/aschlicht/Ceedling/tree/master/plugins/gcov

CezaryGapinski commented:
@aschlicht 

> I didn't know you added the JSON feature to gcvor. Cool! Note that when I used ---json without --output , then no JSON file was created. The gcovr documentation says --json is supported. gcovr --json coverage.json

Thanks! I have to go back to gcovr and try to fix it.

> Back to Ceedling... Your merge idea sounds good. Right now the JSON report is output to build/artifacts/gcov. Does that conflict with your merge idea?

No, I would like to put my temporary files to `build/gcov/out` and next made some operations to generate the final reports to `build/artifacts/gcov`.

> I was thinking that the JSON file input support was to combine coverage information from other portions of project to get a complete picture of coverage results. I didn't think it through though. What if we have a collection of git submodules, each with its own set of unit tests? We combine a set of submodules to create a product. We want to know the unit test coverage for the product, so we specify a set of JSON files to import. Maybe we directly call gcovr for this approach?

The idea is tempting. I'm not sure about paths to the source codes. I guess all gcovr calls must be do to match root directory to correctly merge results (combined options --root and and search_paths) to create JSONs, which could be merged, to correctly display the results in the final report. But I'm not sure.

mvandervoord commented:
Thanks @aschlicht and @CezaryGapinski for your help! This is really great! You've made a lot of developers happy, I'm sure!

jifengtx commented:
> @mvandervoord You're welcome! Thanks for the, "Thank you". I'm glad you appreciate the work. Thank you for promptly chiming in with configuration interface guidelines. I do want to match the look and feel of Ceedling.
> 
> I am happy with the pull request; however, I have a couple more questions. If it makes sense to add Report Generator output to the gcov plugin, then do we want to separate gcovr options from Report Generator options in the project.yml hierarchy?
> 
> 1. Do we want to put all gcovr specific options under a :gcovr: section?
> 2. I think I should rename `:report_types:` to simply `:reports:`. Do you have a better name?
> 
> ```yaml
> :gcov:
>     :report_types: # <--- Rename?
>         - HtmlDetailed
>         - JSON
>     :gcovr:
>         :report_root: "."
>         :html_medium_threshold: 75
>         :keep: true
>     :report_generator:
>         :history_directory: "../history"
>         :verbosity: Info
>         :file_filters:
>             - "-^test_.*"
>             - "+..\/..\/Application\/.*"
> ```

jifengtx commented:
@aschlicht I would like to ask you a question about my source files in another path.No overwrite report was generated using the "ceedling gcov:all" directive. How do I configure this in the.project.YAMl file[
[project.txt](https://github.com/ThrowTheSwitch/Ceedling/files/5628013/project.txt)
](url)

aschlicht commented:
First, please try "ceedling utils:gcov" to generate the coverage report.
Note you may combine multiple ceedling commands, such as "ceedling clobber
gcov:all utils:gcov".

If an empty report is created, then try changing the report root option to
point to your source directory.  I have a similar setup with source in
another directory. See the ceedling gcov plugin readme.md for the option
name and description.

https://github.com/ThrowTheSwitch/Ceedling/blob/master/plugins/gcov/README.md

Does this help?

Aaron

On Wed, Dec 2, 2020, 1:55 AM jifengtx <notifications@github.com> wrote:

> @aschlicht <https://github.com/aschlicht> I would like to ask you a
> question about my source files in another path.No overwrite report was
> generated using the "ceedling gcov:all" directive. How do I configure this
> in the.project.YAMl file[
> project.txt
> <https://github.com/ThrowTheSwitch/Ceedling/files/5628013/project.txt>
> ](url)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ThrowTheSwitch/Ceedling/pull/427#issuecomment-737088419>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD3AUXHJBQ7HKTPLRXS72DDSSX6HTANCNFSM4JONMAEQ>
> .
>
